nomad/state/state_store.go

```diff
@@ -972,18 +969,46 @@ func upsertNodeCSIPlugins(txn *memdb.Txn, node *structs.Node, index uint64) erro
 		return nil
 	}
 
+	inUse := map[string]struct{}{}
 	for _, info := range node.CSIControllerPlugins {
 		err := loop(info)
 		if err != nil {
 			return err
 		}
+		inUse[info.PluginID] = struct{}{}
 	}
 
 	for _, info := range node.CSINodePlugins {
 		err := loop(info)
 		if err != nil {
 			return err
 		}
+		inUse[info.PluginID] = struct{}{}
+	}
+
+	// remove the client node from any plugin that's not
+	// running on it.
+	iter, err := txn.Get("csi_plugins", "id")
+	if err != nil {
+		return fmt.Errorf("csi_plugins lookup failed: %v", err)
+	}
+	for {
+		raw := iter.Next()
+		if raw == nil {
+			break
+		}
+		plug := raw.(*structs.CSIPlugin).Copy()
```

Comment:
I don't think it matters, but we'd save a little garbage waiting to copy until after we know that we want to change the plugin

Reply:
Good call. Done.